cudapoa/include/claraparabricks/genomeworks/cudapoa/utils.hpp

```diff
@@ -50,6 +57,136 @@ void get_multi_batch_sizes(std::vector<BatchSize>& list_of_batch_sizes,
+{
+    std::ifstream infile(filename);
+    if (!infile.good())
+    {
+        throw std::runtime_error("Cannot read file " + filename);
+    }
+    std::string line;
+    int32_t num_sequences = 0;
+    while (std::getline(infile, line))
+    {
+        if (num_sequences == 0)
+        {
+            std::istringstream iss(line);
+            iss >> num_sequences;
+            windows.emplace_back(std::vector<std::string>());
+        }
+        else
+        {
+            windows.back().push_back(line);
+            num_sequences--;
+        }
+    }
+
+    if (total_windows >= 0)
+    {
+        if (get_size(windows) > total_windows)
+        {
+            windows.erase(windows.begin() + total_windows, windows.end());
+        }
+        else if (get_size(windows) < total_windows)
+        {
+            int32_t windows_read = windows.size();
+            while (get_size(windows) != total_windows)
+            {
+                windows.push_back(windows[windows.size() - windows_read]);
+            }
+        }
+
+        assert(windows.size() == total_windows);
+    }
+}
+
+/// \brief Parses windows from 1 or more fasta files
+///
+/// \param[out] windows Reference to vector into which parsed window
+///                     data is saved
+/// \param[in] input_filepaths Reference to vector containing names of fasta files with window data
+/// \param[in] total_windows Limit windows read to total windows, or
+///                          loop over existing windows to fill remaining spots.
+///                          -1 ignored the total_windows arg and uses all windows in the file.
+inline void parse_fasta_windows(std::vector<std::vector<std::string>>& windows, const std::vector<std::string>& input_paths, const int32_t total_windows)
+{
+    const int32_t min_sequence_length = 0;
+    const int32_t num_input_files     = input_paths.size();
+    std::vector<std::shared_ptr<io::FastaParser>> fasta_parser_vec(num_input_files);
+    std::vector<int32_t> num_reads_per_file(num_input_files);
+    for (int32_t i = 0; i < num_input_files; i++)
+    {
+        fasta_parser_vec[i]   = io::create_kseq_fasta_parser(input_paths[i], min_sequence_length, false);
+        num_reads_per_file[i] = fasta_parser_vec[i]->get_num_seqences();
+    }
+    const int32_t num_reads = num_reads_per_file[0];
+
+    for (int32_t i = 1; i < num_input_files; i++)
+    {
+        if (num_reads_per_file[i] != num_reads)
+        {
+            std::cerr << "Failed to read input files." << std::endl;
+            std::cerr << "Number of long-reads per input file do not match with each other." << std::endl;
+            assert(false);
+            return;
+        }
+    }
+    windows.resize(num_reads);
+
+    int32_t idx = 0;
+    for (auto& window : windows)
+    {
+        for (int32_t i = 0; i < num_input_files; i++)
+        {
+            window.push_back(fasta_parser_vec[i]->get_sequence_by_id(idx).seq);
+        }
+        idx++;
+    }
+    if (total_windows >= 0)
```

Comment:
I think this part of the code is shared between the `parse_window_data_file` and `parse_fasta_windows` - let's move into a common function?

Reply:
Moved